### ISSUE TITLE
ENH: KeyError from missing column should list available columns

### ISSUE THREAD

**Issue by janosh** [pandas-dev/pandas#50076]
### Feature Type

- [ ] Adding new functionality to pandas

- [X] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

```py
import pandas as pd

df = pd.util.testing.makeMixedDataFrame()

print(f'{list(df)=}')
>>> list(df)=['A', 'B', 'C', 'D']

df[['foo']]
>>> KeyError: "['foo'] not in index"
```

### Feature Description

The error message would be more helpful if it listed available columns:

df[['foo']]
>>> KeyError: "['foo'] not in columns=['A', 'B', 'C', 'D']"
```

### Alternative Solutions

n/a

### Additional Context

_No response_

**Comment by janosh** [comment]
Could check if 
```py
if len(', '.join(df)) < some_treshold:
    raise KeyError(f"['foo'] not in columns={', '.join(df)}")
```

and make `some_treshold` configurable via, say, `pd.options.key_errors.max_col_list_len`.

**Comment by mroeschke** [comment]
Yeah as is I would be -0.5 to include this given @rhshadrach concerns

1. In an interactive environment, one can quickly access `df.columns` or `df.index` after the error
2. In a script/process, I guess it could be useful in the traceback with error logging infrastructure but may be too verbose more often than not

**Comment by janosh** [comment]
> but may be too verbose more often than not

More often than not column count and name lengths should be manageable, no?

> In a script/process, I guess it could be useful in the traceback with error logging infrastructure

Exactly, that's my use case! When a job fails and I only see it several hours later in a workflow with a dozen different dataframes, it can be hard to determine which data access is failing and how to fix it. I usually have to rerun the script interactively and print column names to determine the fix.

**Comment by kostyafarber** [comment]
Hey I'd like to work on this. Do we want to go ahead with making these changes?

**Comment by phofl** [comment]
This needs more discussion first.

I am also leaning more towards no. We don't want to have a million options

**Comment by jbrockmendel** [comment]
Agreed with other maintainers this isn't the way to go. Closing as no-action.
